File: scriptRunner.py
#!/usr/bin/python
# scriptRunner will run all scripts within a folder that you specify.
# Useful for LaunchAgents and running multiple scripts at user login.
# Thanks to Greg Neagle for an example of how to do this

import optparse
import os
import subprocess
import plistlib
import datetime
import sys
import stat


def main():
    """Main"""

    p = optparse.OptionParser()
    p.set_usage("""Usage: %prog [options]""")
    p.add_option('--once', '-o', dest='runOnce',
                 help="""Directory of scripts to run only once.""")
    p.add_option('--every', '-e', dest='runEvery',
                 help="""Directory of scripts to run every time.""")

    options, arguments = p.parse_args()

    # Check to see if passed options are a directory or not
    for path in (options.runOnce, options.runEvery):
        if path is not None:
            if not os.path.isdir(path):
                sys.exit(path + " is not a directory")

    runOncePlist = os.path.expanduser("~/Library/Preferences/") + "com.company.scriptrunner.plist"

    try:
        runOncePlistValues = plistlib.readPlist(runOncePlist)
    except IOError:
        runOncePlistValues = {}

    if options.runEvery:
        for script in os.listdir(options.runEvery):
            st = os.stat(os.path.join(options.runEvery, script))
            mode = st.st_mode
            if not mode & stat.S_IWOTH:
                try:
                    subprocess.call(os.path.join(options.runEvery, script), stdin=None)
                except OSError:
                    print "Something went wrong!"
            else:
                print script + " is not executable or has bad permissions"

    if options.runOnce:
        for script in os.listdir(options.runOnce):
            if script in runOncePlistValues:
                print os.path.join(options.runOnce, script) + " already run!"
            else:
                st = os.stat(os.path.join(options.runOnce, script))
                mode = st.st_mode
                if not mode & stat.S_IWOTH:
                    try:
                        subprocess.call(os.path.join(options.runOnce, script), stdin=None)
                        runOncePlistValues[script] = datetime.datetime.now()
                    except OSError:
                        print "Uh oh"
                else:
                    print script + " is not executable or has bad permissions"
        plistlib.writePlist(runOncePlistValues, runOncePlist)


if __name__ == '__main__':
    main()

File: scriptRunner.rb
# == Synopsis
#
# scriptRunner: Runs scripts either once or every time
#
# == Usage
#
# scriptRunner.rb -o /path/to/run_once -e /path/to/run_every
#
# -h, --help:
#    show help
#
# --once /path/to/run_once, -o /path/to/run_once
#    Run scripts in 'run_once' one time
#
#
# --every /path/to/run_every, -e /path/to/run_every
#    Run scripts in 'run_every' one time
#

require 'getoptlong'
require 'rdoc/usage'
require 'osx/cocoa'
require 'time'

opts = GetoptLong.new(
  [ '--help', '-h', GetoptLong::NO_ARGUMENT ],
  [ '--once', '-o', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--every', '-e', GetoptLong::OPTIONAL_ARGUMENT ]
)

run_once = nil
run_every = nil

opts.each do |opt, arg|
  case opt
  when '--help'
    RDoc::usage
  when '--once'
    if arg.nil? || arg == '' then
      raise ArgumentError, arg + '--once argument is nil or empty'
    else
      if File.directory? arg then
        run_once = arg
      else
        raise ArgumentError, arg + ' is not a directory'
      end
    end
  when '--every'
    if arg.nil? || arg == '' then
      raise ArgumentError, arg + '--every argument is nil or empty'
    else
      if File.directory? arg then
        run_every = arg
      else
        raise ArgumentError, arg + ' is not a directory'
      end
    end
  end
end



if run_every then
  Dir.foreach(run_every) do |script|
    next if script == '.' or script == '..'
    script_path = File.join(run_every, script)
    if [2, 3, 6 ,7].include?(Integer(sprintf("%o", File.stat(script_path).mode)[-1,1]))
      puts "#{script_path} has dubious permissions"
    elsif !File.executable?(script_path)
      puts "#{script_path} is not executable"
    else
      system(script_path)
    end
  end
end

if run_once then
  run_once_plist = File.expand_path("~/Library/Preferences/" + "com.company.scriptrunner.plist") 
  plist_contents = OSX::NSMutableDictionary.dictionaryWithContentsOfFile(run_once_plist)
  if plist_contents == nil then
    plist_contents = OSX::NSMutableDictionary.alloc.init 
  end
  
  Dir.foreach(run_once) do |script|
    next if script == '.' or script == '..'
      if plist_contents.has_key?(script)
        puts "#{script} has already been run"
      else
        script_path = File.join(run_once, script)
        if [2, 3, 6 ,7].include?(Integer(sprintf("%o", File.stat(script_path).mode)[-1,1]))
          puts "#{script_path} has dubious permissions"
        elsif !File.executable?(script_path)
          puts "#{script_path} is not executable"
        else
          system(script_path)
          plist_contents[script] = Time.now.utc.iso8601 
        end
      end
  end
  plist_contents.writeToFile_atomically(run_once_plist,true)
end
